### ISSUE TITLE
'Bitcoin JSON-RPC: Work queue depth exceeded. Code:429'

### ISSUE THREAD

**Issue by Levino** [bitpay/insight-api#492]
Looks like this happens when many requests arrive in parallel. I would say this needs to be fixed asap. It kind of renders the api useless as it is very easy to ddos...

**Comment by Levino** [comment]
Just do 200 queries in parallel.

**Comment by braydonf** [comment]
Running multiple nodes, likely internal/private, is the means to make it handle more requests.

**Comment by Levino** [comment]
@braydonf Where do you want to continue this?

**Comment by Levino** [comment]
My opinion: It is acceptable for queries to become slow if a lot of queries arrive at the same time, but they should not return errors.

Possible solution: Queue ALL queries to bitcoind on a global level for the running node process. I have implemented a solution with async [here](https://github.com/satoshipay/bitcore-notifications/blob/feature/queue/lib/index.js#L88), which could reused either in bitcore-node or here.

**Comment by Levino** [comment]
So I wrote a script / test that just makes the insight api inaccessible for everyone else. I am reluctant to share it. One just starts it and keeps it running and the api is just down. I am happy to share the script privately. I think this issue should be addressed asap.

**Comment by slush0** [comment]
I second this. Global queue might solve the problem easily.

With current bitcore architecture, even a single user of myTREZOR can DoS whole server just by regular use. This was not an issue in previous architecture, still used by production myTREZOR app, which servers blockhain for thousands users.

**Comment by elichai** [comment]
I have this error too.
I get the `Work queue depth exceeded` Error a lot of times.
Is there any solutions?

**Comment by Levino** [comment]
No. It is not an error, it is so to say what should be expected from the new set up. The api just does not scale in the current version. Try to "come back later" or use your own private servers with the api running.

**Comment by Levino** [comment]
> Global queue might solve the problem easily.

I do not think so. It should be very easy to flood the global queue. One should additionally introduce rate limitation on ip basis or so. Maybe a "queue per ip" could work, where one takes on query from each queue in a circular second queue... This is a clusterf**k because bitcoind just does not scale.

**Comment by elichai** [comment]
I use it on my own server running with bitcore-node.

**Comment by slush0** [comment]
Is performance of newer bitcore-node (using bitcoind over RPC) expected to be much worse than original architecture? If so, are there any plans to keep original bitcore-node maintained (patch for 0.13, ...)?

You're saying that bitcore does not scale well. I cannot second this; we do not have any performance issues so far. The only problem seems to be artifical throttle in form of RPC API in between bitcore and bitcoind.

**Comment by braydonf** [comment]
@slush0 For clarity, the performance in the latest version ([using indexes in bitcoind](https://github.com/bitpay/bitcoin/tree/0.12.1-bitcore)) is many times better than previous. The reads to leveldb match that from the [leveldb benchmarks](https://github.com/google/leveldb#performance). Queries for txids, utxos and balances for addresses should all be improved, as well as the performance while reindexing. The typical bottleneck now is disk speed and LevelDB block compaction.

I would take a look at adjusting the configuration options regarding the work queue limit, timeout and threads available for RPC. And to monitor the `blocks/index/LOG` for activity.
```
rpcworkqueue=<n>
rpcthreads=<n>
rpctimeout=<n>
```

**Comment by Levino** [comment]
Interesting. What are the defaults on these values? What are sane values for these? I cannot find consistent documentation of these values.

About the performance / stability: I challenge you to name me one deployment I can not DOS with zero effort @braydonf

**Comment by matatakun** [comment]
I'm also having the same issue while receiving a new block and querying the api, not while reindexing. Any solutions yet? @braydonf @Levino

**Comment by technyne** [comment]
I'm having the same issues, new installation, can't even get past 3% sync to the network before depth is exceeded. I am going to try implementing some of the changes suggested here and elsewhere but it seems that no one from the dev team is responding to these requests anymore...

**Comment by rachyandco** [comment]
@Levino you can find the default values in [httpserver.h](https://github.com/bitcoin/bitcoin/blob/c0ddd32bf629bb48426b0651def497ca1a78e6b1/src/httpserver.h)

static const int DEFAULT_HTTP_THREADS=4;
static const int DEFAULT_HTTP_WORKQUEUE=16;
static const int DEFAULT_HTTP_SERVER_TIMEOUT=30;

**Comment by runn1ng** [comment]
see PR above

**Comment by Levino** [comment]
As I said in https://github.com/bitpay/bitcoind-rpc/pull/23#issuecomment-357864919 I meanwhile doubt that a simple queue will do anything to help. We need spam protection, so basically a queue per user (for example IP address) with a maximum queue size which returns error 429 (too many requests) when the queue size for this user has been reached.

**Comment by Levino** [comment]
Possible solution:

- Merge the global queue feature in the aforementioned PR.
- Use traefik as a reverse proxy in front of the insight-api with rate limiting http://v1-5.archive.docs.traefik.io/configuration/commons/#rate-limiting

**Comment by winteraz** [comment]
This happens with no request received as well. I just started the service and during the index I've got the same error even if I had all the inbond ports to the API  blocked.

**Comment by charleslcso** [comment]
Dudes, a new install here. Sync'ed to 8% and got this error.... 3 times 24hours of sync'ing... not sure if it's gone.

I guess it's no solution in sight.

BTW, traefik link above seems to be gone... not that I can understand it quickly...

**Comment by Levino** [comment]
Traefik 1.5 has been released. Fixed the link. Has not too much to do with the problem you encounter: You cannot rate limit the requests during sync. Did you try `rpcworkqueue=512` in your `bitcoin.conf`?

**Comment by charleslcso** [comment]
I have no idea what to try.... ha Ha Ha. Thx, will test with your recommendation.
